pkg/koordlet/runtimehooks/hooks/tc/tc_linux.go

```diff
@@ -0,0 +1,745 @@
+	"github.com/koordinator-sh/koordinator/pkg/koordlet/runtimehooks/protocol"
+	"github.com/koordinator-sh/koordinator/pkg/koordlet/runtimehooks/reconciler"
+	"github.com/koordinator-sh/koordinator/pkg/koordlet/runtimehooks/rule"
+	"github.com/koordinator-sh/koordinator/pkg/koordlet/statesinformer"
+	"github.com/koordinator-sh/koordinator/pkg/koordlet/util/system"
+	sysutil "github.com/koordinator-sh/koordinator/pkg/koordlet/util/system"
+)
+
+const (
+	name        = "tcPlugin"
+	description = "setup tc rules for node"
+
+	ruleNameForNodeSLO = name + " (nodeSLO)"
+	ruleNameForAllPods = name + " (allPods)"
+)
+
+const (
+	MAJOR_ID              = 1
+	QDISC_MINOR_ID        = 0
+	ROOT_CLASS_MINOR_ID   = 1
+	SYSTEM_CLASS_MINOR_ID = 2
+	LS_CLASS_MINOR_ID     = 3
+	BE_CLASS_MINOR_ID     = 4
+
+	// 0-7, In the round-robin process, classes with the lowest priority field are tried for packets first.
+	SYSTEM_CLASS_PRIO = 0
+	LS_CLASS_PRIO     = 1
+	BE_CLASS_PRIO     = 2
+
+	// Maximum rate this class and all its children are guaranteed. Mandatory.
+	// attention: the values below only represent the percentage of bandwidth can be used by different tc classes on the host,
+	// the real values need to be calculated based on the physical network bandwidth.
+	// eg: eth0: speed:200Mbit => high_clss.rate = 200Mbit * 40 / 100 = 80Mbit
+	BE_CLASS_RATE_PERCENTAGE     = 30
+	LS_CLASS_RATE_PERCENTAGE     = 30
+	SYSTEM_CLASS_RATE_PERCENTAGE = 40
+
+	// Maximum rate at which a class can send, if its parent has bandwidth to spare.  Defaults to the configured rate,
+	// which implies no borrowing
+	CEIL_PERCENTAGE = 100
+
+	DEFAULT_INTERFACE_NAME = "eth0"
+)
+
+var (
+	rootClass   = netlink.MakeHandle(MAJOR_ID, ROOT_CLASS_MINOR_ID)
+	systemClass = netlink.MakeHandle(MAJOR_ID, SYSTEM_CLASS_MINOR_ID)
+	lsClass     = netlink.MakeHandle(MAJOR_ID, LS_CLASS_MINOR_ID)
+	beClass     = netlink.MakeHandle(MAJOR_ID, BE_CLASS_MINOR_ID)
+
+	ipsets = []string{string(NETQoSSystem), string(NETQoSLS), string(NETQoSBE)}
+)
+
+type tcPlugin struct {
+	rule *tcRule
+
+	allPodsSyncOnce sync.Once // sync once for AllPods
+
+	// this is the physical NIC on host, default eth0
+	interfLink netlink.Link
+
+	// for executing the iptables command.
+	iptablesHandler *iptables.IPTables
+	// for executing the tc and ipset command.
+	netLinkHandler netlink.Handle
+
+	executor resourceexecutor.ResourceUpdateExecutor
+}
+
+var singleton *tcPlugin
+
+func Object() *tcPlugin {
+	if singleton == nil {
+		singleton = newPlugin()
+	}
+	return singleton
+}
+
+func newPlugin() *tcPlugin {
+	return &tcPlugin{
+		netLinkHandler:  netlink.Handle{},
+		allPodsSyncOnce: sync.Once{},
+		rule:            newRule(),
+	}
+}
+
+func (p *tcPlugin) Register(op hooks.Options) {
+	klog.V(5).Infof("register hook %v", name)
+
+	rule.Register(ruleNameForNodeSLO, description,
+		rule.WithParseFunc(statesinformer.RegisterTypeNodeSLOSpec, p.parseRuleForNodeSLO),
+		rule.WithUpdateCallback(p.ruleUpdateCb),
```

Comment:
maybe only need prepare for nodeSLO callback?

Reply:
ipset need to update from pod.